pkg/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,385 @@
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) syncGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	var err error
+	func() {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		// only call List if we are still watching the gvk.
+		if c.watchedSet.Contains(gvk) {
+			err = c.reader.List(ctx, u)
+		}
+	}()
+
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
+	}
+
+	for i := range u.Items {
+		if err := c.AddObject(ctx, &u.Items[i]); err != nil {
+			return fmt.Errorf("adding data for %+v: %w", gvk, err)
+		}
+	}
+
+	return nil
+}
+
+func (c *CacheManager) manageCache(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-c.backgroundManagementTicker.C:
+			func() {
+				c.mu.Lock()
+				defer c.mu.Unlock()
+
+				c.wipeCacheIfNeeded(ctx)
+
+				// spin up new goroutines to relist gvks as there has been a wipe
+				if c.needToList {
+					// stop any goroutines that were relisting before
+					// as we may no longer be interested in those gvks
+					close(c.relistStopChan)
+
+					// assume all gvks need to be relisted
+					gvksToRelist := c.gvksToSync.GVKs()
+
+					// clean state
+					c.needToList = false
+					c.relistStopChan = make(chan struct{})
+
+					go c.replayGVKs(ctx, gvksToRelist)
+				}
+			}()
+		}
+	}
+}
+
+func (c *CacheManager) replayGVKs(ctx context.Context, gvksToRelist []schema.GroupVersionKind) {
+	gvksSet := watch.NewSet()
+	gvksSet.Add(gvksToRelist...)
+
+	for gvksSet.Size() != 0 {
+		gvkItems := gvksSet.Items()
+
+		for _, gvk := range gvkItems {
+			select {
+			case <-ctx.Done():
+				return
+			case <-c.relistStopChan:
+				return
+			default:
+				operation := func() (bool, error) {
+					if err := c.syncGVK(ctx, gvk); err != nil {
+						return false, err
+					}
+					return true, nil
+				}
+
+				if err := wait.ExponentialBackoff(backoff, operation); err != nil {
+					log.Error(err, "internal: error listings gvk cache data", "gvk", gvk)
+				} else {
+					gvksSet.Remove(gvk)
+				}
+			}
+		}
+
+		c.ReportSyncMetrics()
+	}
+}
+
+// wipeCacheIfNeeded performs a cache wipe if there are any gvks needing to be removed
+// from the cache or if the excluder has changed. It also marks which gvks need to be
+// re listed again in the cf data cache after the wipe.
+// assumes the caller has lock.
+func (c *CacheManager) wipeCacheIfNeeded(ctx context.Context) {
+	// remove any gvks not needing to be synced anymore
+	// or re evaluate all if the excluder changed.
+	if c.gvksToDeleteFromCache.Size() > 0 || c.excluderChanged {
+		if err := c.wipeData(ctx); err != nil {
```

Comment:
instead of if/else you can just do an early return and save yourself some nesting:

```
	// remove any gvks not needing to be synced anymore
	// or re evaluate all if the excluder changed.
	if c.gvksToDeleteFromCache.Size() > 0 || c.excluderChanged {
		if err := c.wipeData(ctx); err != nil {
			log.Error(err, "internal: error wiping cache")
			return
		}

		c.gvksToDeleteFromCache = watch.NewSet()
		c.excluderChanged = false
		c.needToList = true
	}
```

Reply:
This latest code should reflect the proposed changes.